sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/CoalesceShufflePartitions.scala

```diff
@@ -146,13 +147,15 @@ case class CoalesceShufflePartitions(session: SparkSession) extends AQEShuffleRe
       Seq(collectShuffleStageInfos(r))
     case unary: UnaryExecNode => collectCoalesceGroups(unary.child)
     case union: UnionExec => union.children.flatMap(collectCoalesceGroups)
-    // If not all leaf nodes are exchange query stages, it's not safe to reduce the number of
-    // shuffle partitions, because we may break the assumption that all children of a spark plan
-    // have same number of output partitions.
     // Note that, `BroadcastQueryStageExec` is a valid case:
     // If a join has been optimized from shuffled join to broadcast join, then the one side is
     // `BroadcastQueryStageExec` and other side is `ShuffleQueryStageExec`. It can coalesce the
     // shuffle side as we do not expect broadcast exchange has same partition number.
+    case join: BroadcastHashJoinExec => join.children.flatMap(collectCoalesceGroups)
+    case join: BroadcastNestedLoopJoinExec => join.children.flatMap(collectCoalesceGroups)
+    // If not all leaf nodes are exchange query stages, it's not safe to reduce the number of
+    // shuffle partitions, because we may break the assumption that all children of a spark plan
+    // have same number of output partitions.
     case p if p.collectLeaves().forall(_.isInstanceOf[ExchangeQueryStageExec]) =>
```

Comment:
so we were fine before for simple queries because this condition filters out the broadcast joins?

Reply:
Can we reproduce the bug without union?

Reply:
It seems we should improve this line to check the if the child of `BroadcastQueryStageExec` exists `Union`.

Reply:
> so we were fine before for simple queries because this condition filters out the broadcast joins?

No, this condition won't filter out the broadcast joins, we filter the right side for broadcast join in `collectShuffleStageInfos`

Reply:
> It seems we should improve this line to check if the child of `BroadcastQueryStageExec` exists `Union`.

In fact, I have tested the queries with various join types and join types with 'union', and found that `ShuffledHashJoin` and `SortMergeJoin` work normally, while `BroadcastJoin` and `CartesianProduct` with `Union` fail. The former case we should ignore broadcast joins, while the latter requires additional detection. Maybe we should add the union check and solve the latter cases in this pr?

Reply:
> we filter the right side for broadcast join in collectShuffleStageInfos

I didn't see any code related to it, can you elaborate?

Reply:
When [collectShuffleStageInfos](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/CoalesceShufflePartitions.scala#L157), called in [collectCoalesceGroups](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/CoalesceShufflePartitions.scala#L144), collect the shuffleStageInfos of the subplans, it will only traverse `ShuffleQueryStageExec` and `AQEShuffleReadExec`, the `BroadcastQueryStageExec` will be excluded. Therefore, when a SQL only has broadcastjoin, only the valid shuffleStage can be obtained.